docs/design_notes/lift.md

```diff
@@ -0,0 +1,340 @@
+
+A number of decisions had to be made in order to define `pack`. The implementation
+of `pack` controls how variables and rngs are lifted and how fine-grained the user control is.
+It must also decide whether lifting decisions are made at variable or transformation definition.
+
+
+### Lifting granularity
+
+
+With the Linen API, users can define arbitrary variable collections and PRNG sequences.
+Each variable in a collection is lifted in the same way.
+
+Collections are typically given a semantically meaningful name like "params" or "batch_stats" rather than a general purpose name like "state".
+Because collections carry semantic meaning we can decide at the transformation level how each collection should be lifted.
+For example, we want to share all parameter variables when we add a batch dimension to a model.
+
+At the same time we can write generic code that uses transformations without knowing exactly what kind of variables the submodules will create.
+Collections thus strike a balance between fine-grained control and generality.
+We also avoid brittle string matching code that loops over all variables and tries to split up collections in an ad-hoc way based on
+naming conventions like: target all variables with the name prefix "kernel". 
+If more fine-grained control is necessary a user can simply split up a set of variables over multiple collections that should be handled differently.
+
+
+### Transformation vs variable control
+
+
+Lifting behavior could be defined either at the transformation level or during variable definition.
+We use transformation level definitions of lifting behavior.
+The reason for this choice is that there are many different transformations with various behaviors.
+For example: `vmap` has broadcasted and vectorized arguments, while `scan` has scan, carry, and broadcast arguments.
+A variable would have to define its behavior for all these transformations otherwise a `Module` would not be compatible with
+these transformations. Alternatively, we would have to make default decisions for how transformations are handled.
+However, this could lead to silent bugs because the behavior might not actually be valid given the users intent. 
+
+The lift package also provides a general purpose `transform`, which allows an arbitrary function to transform a variable collection.
+For example, this can be used to tie the weights in a tied auto-encoder by transposing the weights.
+It is unclear whether a similar general purpose transform could be defined if lifting decisions were made at variable definition.
+
+
+### Linen
+
+The lifting module does not know about the Linen `Module` API. Instead it operates directly on instances of `flax.core.Scope`.
+A `Scope` instance contains the variables and PRNG sequences of a `Module`.
+Each module has a `Scope` instance linked if it is bound to a parent or it was created using `init` or `apply`.
```

Comment:
This sentence is a bit hard to parse:
- What does "has a scope instance linked" mean exactly? (e.g. what's the alternative)
- It's not clear if it's "(has a scope instance linked if it is bound to a parent) or (it was created using init/apply)" vs. "has a scope instance linked (if it is bound to a parent) or (it was created using init/apply)"

Reply:
rephrased